### ISSUE TITLE
[DISCUSSION] Fluid Engine - Next Steps, Branches and Major Releases

### ISSUE THREAD

**Issue by bmack** [TYPO3/Fluid#561]
Hey maintainers,

currently we have a situation where it's unclear how this project will continue, as far as I see from the outside perspective, and we're in maintenance mode.

* There is a "master" branch, which reflects the current 2.x development series with a few commits in the past 12 months
* There is a "3.x-dev" branch, containing last changes from June 2019
* There is a "3.0" branch with serious refactoring related to the actual Template Parser, but also the template handling and new concepts such as Components and Atoms instead of Partials, Layouts and Templates - last change was 12 months ago.

TYPO3 Core is utilizing the current 2.x development series.

For this reason, I propose the following approach forward:

1. Branch off "2.7" from master (as we have done with 2.6 etc)

2. Release a 3.0 version with a new Template Parser, PHP 7.4+-only support and no other substantial changes, ensuring that TYPO3 Core v11 / v12 can update to the next major version without any hard migrations. This work could be done within the master branch. We could create a Github "Project" in this repository to summarize all the things that will be breaking (I guess mostly internally) when updating from 2.7 to 3.0. I propose to start off working with the 2.x development series (= master branch) here instead of using the "3.x-dev" or "3.0" branches as the 2.x development series contains the important security fix (CVE-2020-26216) including tests.

3. Add functionality around Components after the 3.0 release as minor feature releases without having to add breaking changes, allowing to switch between the Atoms approach and the Partials concept. If breaking changes are necessary, we could also release a 4.0 release. For this concept to be included, I suggest to find a good way on how to document and how to migrate to this new concept (this is still unclear to me what the details mean for template builders or users, or VH users).

4. Create a new version with deprecations of all the old concepts. This way 3.x is a bridge between the old concepts and the new ones. A new major version then would have a clear and clean code base. This new major version will approx. be in 12 months or so.

Side note:
Since 3.0 has been "hanging" for such a long time, I would recommend an easy upgrade path to 3.0. Alternatively, we could even skip "3.0" and go with "4.0" directly in order to avoid confusion for people already trying out the features of the existing "3.0" branch.

Thanks for consideration.
Benni.

**Comment by ohader** [comment]
I like the proposed approach, as it helps to make version `3.0` (or maybe `4.0` as outlined in the end) official.

However, I'd like to outline that the XSS vulnerability as addressed in `CVE-2020-26216` is probably not fixed in corresponding `3.0` or `3.x-dev` branches. This should be verified and addressed before creating official releases of those branches. Please see https://typo3.org/security/advisory/typo3-core-sa-2020-009 for details.

**Comment by NamelessCoder** [comment]
Although I would of course like to proceed here, the sketched plan unfortunately wouldn't work since the rewritten TemplateParser is completely interdependent on the Components concept (and basically all other new features with a few minor exceptions to that rule). Due to the nature of the changes I don't see a feasible path to a partial new major - I'm afraid it's an "all or nothing" situation. The logical next major version should then be 3.0 and it should contain all the changes you see in the `3.0` branch - otherwise IMHO it doesn't make much sense to create a new major version (as there would be no breaking changes).

We could however create an intermediate 2.x release which deprecates the parts that will be removed or significantly changed in the future. This has not been considered up until now, but I like the idea.

I'm all for creating a "Project" to keep track of the incompatibilities! In terms of a hard migration - every effort was made to avoid having to do literally any migration at all, except for some cases of advanced integrations (e.g. custom compiling methods or implementations based on making modifications to parsed nodes or arguments). That said, there are still some bugs that mean 3.0 behaves different from 2.x, but the aim is to solve those (see recent PRs from Sascha Egerer which addresses all of the bugs Sascha encountered while testing 3.0 against TYPO3 CMS with some actual real-world projects).

By the way, the intention was that projects implementing Fluid wouldn't necessarily have to force use of 3.0 but could implement Fluid in a way that allows either 2.x or 3.x and then, for example, let the third party packages constrain to either 2.x or 3.x based on the third party packages' compatibility with those versions. Ideally though, third party packages should if necessary be updated to support 3.x API - but this would only be strictly necessary in case of very complex or low-level integration (for example, the VHS extension's ViewHelpers are actually compatible with both versions except for a few rough edges that can easily be taken care of by detecting if 2.x or 3.x is being used).

I should also note here that the plan is not to switch between Partials and Atoms concepts but rather, allow both to coexist for as much time as we deem necessary, and slowly deprecate the Partials approach - which then of course wouldn't be dropped until 4.0 or even further into the future. The backwards compatible support for Partials (and Layouts) is quite small and in my evaluation does not pose any risk for incompatibility with the Atoms concept.

`CVE-2020-26216` indeed has to be addressed in the `3.0` branch as well. Luckily, the concept of argument escaping that was introduced as solution to that vulnerability should be possible to copy 1:1 to the `3.0` branch (but probably not as a simple cherry pick due to the fact that argument parsing is completely rewritten and is much more integrated in the TemplateParser in 3.0).

In short:

* I would not make an intermediate major version since this doesn't make technical sense (none of the smaller parts can be separated from the whole and if we separate enough parts to justify a new major, then this release would inevitably be precisely as breaking as the already planned 3.0 major).
* I would instead opt for taking care of obviously the security issue and additionally all the bugs discovered by Sascha, which would result in an essentially migration-free upgrade path (except for some advanced or very low-level implementations; of which there are none in the TYPO3 CMS core itself).

**Comment by bmack** [comment]
I see your point, however I fear it's not as trivial as this looks:

Some Scenarios for TYPO3:

1. Extension A is not compatible with TYPO3 Core, but TYPO3 Core ships with 3.x, Extension B is using 3.0-only features

2. A site owner wants to use Fluid 3.x features but Core ships 2.x for legacy reasons (non-composer) - no way for Core to distinguish between different template versions for v2 and v3.

3. An extension explicitly ensures Fluid 2.x is used, but the other extension is restricting to Fluid 3.x, both won't work together in a single installation.

A global (= installation-wide) switch in TYPO3 Core does not seem feasable for me to switch here.

As for an Upgrade Guide, the only official document I've found so far is this news: https://typo3.org/article/fluid-30-workshop-summary-important-updates-for-users - or is there anything else?

In addition (as far as I understood the concepts and transformation), I tend to switch to use just "Templates" instead of "Atoms" as a naming convention. In Fluid 3.0 means - any Template file can be a partial or extended or included -  Templates can be included and inherited, getting rid of the Partials and Layout concept. But maybe I did not understand the Atoms concept fully?

**Comment by NamelessCoder** [comment]
Let's take the easy part first - the link you found to the t3org article does describe the main incompatibilities but it is also a bit brief. It would make a nice starting point for the "Project" to keep track of major changes. If you need more detailed info there's an example project https://github.com/NamelessCoder/fluid-3-demo which if you check it out locally and run the php webserver can be browsed and has source examples for all the new stuff.

I do see your point with compatibility in non-composer mode (I tend to forget about non-composer mode still existing and I wish we would get rid of that, but that's another topic altogether). This leaves us with just two options:

1. We can ship two non-composer dists, one with Fluid 2.x and another with 3 (but of course this also has problems in which extensions it would support and not being able to mix and match).
2. Or we can choose 3.0 as the lowest common denominator (which then may require some extensions to update to achieve compatibility and in worst case, may make it very hard for such extensions to keep compatibility with earlier TYPO3 versions).

Given those two choices I'd lean towards option 2, the impact of which would be much lessened by taking care of the bugs Sascha reported. So that seems like a reasonable option. I only wish that we didn't have to make this choice because non-composer mode is still a thing ;)

Re: naming of Templates vs. Atoms - I think it makes a lot of sense to distinguish between these. Templates has an implied meaning from Fluid 2.x and is imho too generic when searching etc., so I would stay away from that naming. Atoms on the other hand also has an implied meaning, but is far less generic and ensures that there's no confusion (is this "Template" actually an Atom or is it a controller action Template? and so on). In terms of technical differences there *are* some significant differences between what a "Template" is in Fluid 2.x context and what an "Atom" is in Fluid 3.0 context - but the differences if you're just looking af Fluid 3.0 are somewhat negated by the fact that Fluid 3.0 is capable of treating *every* template file as an Atom and every template file has the same capabilities. Basically, any old "Template" is also an "Atom" (it only adds a hint that it's designed to work in a legacy context of Templates, Partials and Layouts but in every other way, it's an Atom). BTW the demo project I linked to goes into a lot of detail about how an Atom is different from e.g. a Partial when comparing Fluid 2.x and 3.0.

**Comment by bmack** [comment]
Thanks, Claus, for providing the examples. The demo is a good place for people to understand the new concepts, indeed.

I had a closer look at the Sequencer, the Parser, the ViewHelpers (f:extend etc) in combination with the demo. I also understand your reasoning on why it would be hard to extract the sequencer itself.

Some remarks which I have by reading the source code:

1. Everything is a Component now. While I see the benefits of it, the ComponentInterface is rather large (15 methods to implement, https://github.com/TYPO3/Fluid/blob/3.0/src/Component/ComponentInterface.php), even larger than the current NodeInterface. I think a separation into various Interfaces (one with Children, one without, not every component has arguments) allows for better Composition, and clarification what is responsible for what. This also would allow us to introduce new functionality or deprecate old functionality in a easier way in the future.


2. The ComponentInterface probably also the reason why AbstractComponent is introduced. The "abstract class" AbstractComponent does not have any abstract methods - because it sets a best practice (https://github.com/TYPO3/Fluid/blob/3.0/src/Component/AbstractComponent.php#L25). Although we've done this in the past in TYPO3, I think this is a rather bad design pattern. It adds a layer of inheritance on top everything that is now extensible in Fluid.


3. The Splitter works with the null-byte sequence as end-of-template. Although this is common around C-based projects, we have to consider invalid bytes here as well (we had some nasty security issues related to the null-byte sequences in the past). https://github.com/TYPO3/Fluid/blob/3.0/src/Core/Parser/Splitter.php#L94 


4. We're still lacking a grammar for the Fluid syntax, which would be beneficial for linting purposes. See https://github.com/nikic/PHP-Parser/blob/master/grammar/php7.y as an example.


5. By mixing the Parser / Sequencer with the new functionality of Atoms, it's now visible that ViewHelpers such as <f:extend> and <f:parameter> are actually not just ViewHelpers but part of the parsing process. https://github.com/TYPO3/Fluid/blob/3.0/src/Core/Parser/SyntaxTree/EntryNode.php#L65-L85
This is probably one of the reasons, why it is hard to decouple the Sequencer from the rest of the functionality.


6. Atoms are much more than templates, I get that. But they then the ViewHelperResolver is used to resolve an atom, mixing ViewHelpers again with Atoms (and with Partials and Templates, as you already mentioned). https://github.com/TYPO3/Fluid/blob/3.0/src/ViewHelpers/ExtendViewHelper.php#L56-L57 - seems like Atoms are the super-power that really make layouts, partials, templates and ViewHelpers obsolete, but they are added in a way as a new concept that intermingles with all the existing parts, without a clear separation of concerns for any of these existing Core components of Fluid.

Especially the last point scares me of Fluid 3.0, but that's just my personal gut feeling.

Benni.

**Comment by mbrodala** [comment]
All links above as permalinks for the future:

* https://github.com/TYPO3/Fluid/blob/e1245ed6bc6114a4e52c18d4db909acac3c00d25/src/Component/ComponentInterface.php
* https://github.com/TYPO3/Fluid/blob/e1245ed6bc6114a4e52c18d4db909acac3c00d25/src/Component/AbstractComponent.php#L25
* https://github.com/TYPO3/Fluid/blob/e1245ed6bc6114a4e52c18d4db909acac3c00d25/src/Core/Parser/Splitter.php#L94
* https://github.com/nikic/PHP-Parser/blob/50953a2691a922aa1769461637869a0a2faa3f53/grammar/php7.y
* https://github.com/TYPO3/Fluid/blob/e1245ed6bc6114a4e52c18d4db909acac3c00d25/src/Core/Parser/SyntaxTree/EntryNode.php#L65-L85
* https://github.com/TYPO3/Fluid/blob/e1245ed6bc6114a4e52c18d4db909acac3c00d25/src/ViewHelpers/ExtendViewHelper.php#L56-L57

Thank me when you come back after a few years. ;-)

**Comment by NamelessCoder** [comment]
@benni Just some quick remarks for your points above:

1. True, the ComponentInterface is rather large but I see two also-large hurdles in the way of dissecting it into smaller interfaces. First, a lot of business logic in for example the Sequencer returns ComponentInterface (and cannot return a smaller interface and still be working; some methods can, but very far from all). Second, if the methods started returning a more specific interface then that interface would probably have to extend ComponentInterface and we wouldn't have gained anything from dissecting them. That said, I'd be happy to take another look with fresh eyes and see what might be possible to do. I should also note that part of the reason why the interface is this large is to keep it compatible with the old NodeInterface to minimise the need for third party implementations that use NodeInterface to make huge changes. This then also implies that if we do dissect the interface, we probably still have to keep the methods that also exist on the old NodeInterface, as part of the main interface. Otherwise we introduce breakage.
2. AbstractComponent is indeed meant like an "I don't want to manually implement all the interface methods" helper - it has no explicitly required functionality and it is fully optional to use or not.  I do think it helps more than it hurts, since even if the ComponentInterface would be dissected, the alternative to the abstract class would be Traits implementing this and at least I personally dislike Traits much more than a common base class.
3. Could you point me to the issues that were previously discovered with null bytes? The reason why the Sequencer supports them is to support a streaming approach where a specific byte is the way to say "here ends the stream of bytes" and for that, null byte termination seems like the go-to solution. It could however be literally any (non-multi) byte so there wouldn't be any technical limits preventing us from choosing another. But I'd like to have a look at the old problems to know if they would be relevant (I don't think they are, because nothing forces a Fluid template to contain a null byte and a premature null byte encountered in a template simply means the sequencing stops early; it's not vulnerable to injection afaics). Null bytes contained within variables of course don't matter at all.
4. A (complete) grammar would be quite difficult to establish since parts of the grammar is dynamic and gets established by whichever implementation handles it - for example, expressions within curly braces. Basically, the behavior or a linter in part depends on how you configured the context for Fluid (e.g. third party components can take over the sequencing and react to literally anything or completely change the normal sequencing, see for example https://github.com/NamelessCoder/fluid-3-demo/blob/master/src/ViewHelpers/PreViewHelper.php which disables all Fluid handling until the closing tag is encountered). I don't see how a grammer could be written that takes such things into account, but maybe I'm overlooking something? Side note, the XML aspect of Fluid can of course still be linted with any standard XML-linter (though it might have to disable namespaced tag schema validation).
5. That is definitely one of the major reasons, along with the fact that how a component is configured (arguments, escaping etc.) also has an effect on how the sequencer processes the tag/attributes/children. This is both the strength and limitation of the sequencer: because it operates directly on components it does not need to first extract string representations and then process those into objects, it all happens on the fly. With increased speed and decreased memory usage as a result, but with increased coupling as a drawback.
6. True, but I guess part of the "confusion" comes from not having renamed the ViewHelperResolver to e.g. ComponentResolver which is in turn a compatibility measure. What ViewHelpers and Atoms have in common is that they're both just types of components and are completely interchangable. They're called with the same syntax and share a common interface. All of this combined is the reason why ViewHelperResolver was given the API to resolve an Atom as well as ViewHelpers. It just seemed more practical than having two resolvers or renaming the ViewHelperResolver and introduce breakage.

**Comment by NamelessCoder** [comment]
> If we have to break the templates, we can and should also discuss completely different templating solutions.

But this is the misconception I'm trying to clear up - we do not have to break the templates. The premise for starting that discussion (here, at least) therefore does not apply.

But I would suggest that TYPO3 CMS has this discussion internally *before* we continue here. Maybe also ask the user base what they'd prefer. If the outcome of that is that Fluid will no longer be used then it seems irrelevant to go into further discussions about what can/should be done with Fluid 3.0.

I request that this decision gets made first - and if the outcome is to continue to use Fluid then I would be more than happy to go through the code base in detail with anyone who's interested and address whichever concerns you may have about the design.

> Fluid 3.0 has been in development for three years

And I wish this hadn't been the case, but a very big part of the reason for this is the lack of reaction from TYPO3 CMS in testing the solution and providing feedback like the points raised by people in this thread. I'm happy that this is happening now, but I don't think it's fair to raise the "long development" point as a negative (because this isn't any fault of myself or other Fluid maintainers).

> Conceptual flaws were fixed with a new parser

This is simply because the conceptual flaws were caused by the parser's usage of regular expressions to do the parsing.

> why not just extract this and finish this one-off?

I've already answered this point earlier in the thread.
